File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int sum = 0;
        int count =0;

        HashMap<Integer, Integer> mp = new HashMap<>();
        mp.put(0, 1); // (sum, its repetition)

        for (int num : nums) {
            sum += num;

            // Check if there is a subarray with a sum of k ending at the current index
            if (mp.containsKey(sum - k)) {
                count += mp.get(sum - k);
            }

            mp.put(sum, mp.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        running_sum_freq_map = {0 : 1} # One way to get Running Sum 0, Empty SubArr
        running_sum = 0
        ans = 0

        for i in range(n):
            running_sum += nums[i]

            # For Current Running Sum, Required Sum will be Current_Running_Sum - k
            required_sum = running_sum - k

            # If required_sum is already Appeared
            if required_sum in running_sum_freq_map:
                # We have to update ans with number of times that required_sum Appeared
                ans += running_sum_freq_map[required_sum]

            # Do Append/Update Current_Running_Sum in Map
            if running_sum in running_sum_freq_map:
                # If Already Present, then Do increment its Count/Freq
                running_sum_freq_map[running_sum] += 1
            else:
                 running_sum_freq_map[running_sum] = 1
                    
        return ans
        